napari/_vispy/canvas.py

```diff
@@ -1,36 +1,90 @@
 
+from napari._vispy import VispyCamera
 from napari._vispy.utils.gl import get_max_texture_sizes
+from napari.utils._proxies import ReadOnlyWrapper
 from napari.utils.colormaps.standardize_color import transform_color
+from napari.utils.interactions import (
+    mouse_double_click_callbacks,
+    mouse_move_callbacks,
+    mouse_press_callbacks,
+    mouse_release_callbacks,
+    mouse_wheel_callbacks,
+)
+
+if TYPE_CHECKING:
+    from napari.components import ViewerModel
 
 
-class VispyCanvas(SceneCanvas):
+class SceneCanvas(SceneCanvas_):
+    """Vispy SceneCanvas used to allow for ignoring mouse wheel events with modifiers."""
+
+    def _process_mouse_event(self, event):
+        """Ignore mouse wheel events which have modifiers."""
+        if event.type == 'mouse_wheel' and len(event.modifiers) > 0:
+            return
+        super()._process_mouse_event(event)
+
+
+class VispyCanvas:
     """SceneCanvas for our QtViewer class.
 
     Add two features to SceneCanvas. Ignore mousewheel events with
     modifiers, and get the max texture size in __init__().
 
+    Parameters
+    ----------
+    viewer : napari.components.ViewerModel
+        Napari viewer containing the rendered scene, layers, and controls.
+
     Attributes
     ----------
     max_texture_sizes : Tuple[int, int]
         The max textures sizes as a (2d, 3d) tuple.
-
+    last_theme_color : Optional[npt.NDArray[np.float]]
+        Theme color represented as numpy ndarray of shape (4,) before theme change
+        was applied.
+    background_color_override : Optional[npt.NDArray[np.float]]
+        Background color of the canvas represented as numpy ndarray of shape (4,) which overrides the
+        canvas background color indicated in the napari theme settings.
+    viewer : napari.components.ViewerModel
+        Napari viewer containing the rendered scene, layers, and controls.
+    scene_canvas : vispy.scene.SceneCanvas
+        The vispy SceneCanvas which automatically draws the contents of a scene.
+    view : vispy.scene.widgets.viewbox.ViewBox
+        Rectangular widget in which a subscene is rendered.
+    vispy_camera : napari._vispy.VispyCamera
+        The camera class which contains both the 2d and 3d camera used to describe the perspective by which a
+        scene is viewed and interacted with.
+    layer_to_visual : dict(napari.layers, napari._vispy.layers)
+        A mapping of the napari layers that have been added to the viewer and their corresponding vispy counterparts.
     """
 
     _instances = WeakSet()
 
-    def __init__(self, *args, **kwargs) -> None:
+    def __init__(self, viewer: ViewerModel, *args, **kwargs) -> None:
 
         # Since the base class is frozen we must create this attribute
         # before calling super().__init__().
         self.max_texture_sizes = None
         self._last_theme_color = None
         self._background_color_override = None
-        super().__init__(*args, **kwargs)
+        self.viewer = viewer
```

Comment:
Comment: in general, I get a little concerned when classes store a reference to `ViewerModel` since that gives them a lot of power to cause side-effects by modifying that model. It seems like we mostly read from this, we makes me a little happier. I don't think things are much worse than they are on `main` and I think I see some ways to improve this in the future (e.g. by adding a `Canvas` model that might absorb some/all of the `ViewerModel` access here.

Reply:
Agree, and the canvas model is definitely the next step after this refactor for that reason as well :)

Reply:
One reason for not having a canvas model at the moment is that this would be an eventedmodel within the Napari API. This would not be a pure refactor but like @brisvag said is indeed a next step.